sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -291,18 +291,18 @@ void Scheduler::removeMemoryObject(detail::SYCLMemObjI *MemObj) {
 
 EventImplPtr Scheduler::addHostAccessor(Requirement *Req) {
   std::vector<Command *> AuxiliaryCmds;
-  Command *NewCmd = nullptr;
+  EventImplPtr NewCmdEvent = nullptr;
 
   {
     WriteLockT Lock(MGraphLock, std::defer_lock);
     acquireWriteLock(Lock);
 
-    NewCmd = MGraphBuilder.addHostAccessor(Req, AuxiliaryCmds);
+    Command *NewCmd = MGraphBuilder.addHostAccessor(Req, AuxiliaryCmds);
+    if (!NewCmd)
+      return nullptr;
+    NewCmdEvent = NewCmd->getEvent();
   }
 
-  if (!NewCmd)
```

Comment:
Is this premature exit the original problem?

Reply:
no, original problem was that addHostAccessor (update requirement cmd and empty command creation) - the code under write lock here may be called by multiple threads (cmd1 for thread1 and command2 for thread2) and cmd1 may be removed from leaves and deleted under lock for cmd2 while code under read lock here that uses this raw pointer to cmd1 is not executed yet (enqueue is done but in in releaseHostAccessor in another thread that enqueue unblocked commands). In this case here we have not null but not valid pointer. So we receive - seg fault (rarely) or hang (frequently) because that memory block is reused for new cmd3 and is dependent on empty cmd for cmd1 that leads to closed dependency.